Migrating flare to a component (#14234)
DataDog/datadog-agent: docs/components/migrating-to-components.md

# Migrating to Components

> Once your component has been created you might have to "link" it to other component such as flare, status page, health,
> ...

Reviewer: ```suggestion
After your component has been created you can link it to other components such as flares, status pages, or health.
```

> This page document what needs to be done for each of them so you component is fully integrated in the Agent life cycle

Reviewer: ```suggestion
This page documents how to fully integrate your component in the Agent life cycle.
```

## Flare

Migrate the code related to your component's domain from `pkg/flare` to your component. After migrating

> the Agent to components, you can delete `pkg/flare`.

Author: I think each team should own their code in `pkg/flare`. Only they would know when a piece of code has been correctly migrated and can be deleted. I'll rewrite `you can delete 'pkg/flare'` to `you can delete it from 'pkg/flare'` to be a bit clearer.

> You need to migrate the code related to your component's domain from `pkg/flare` to your component. Once the migration

Reviewer: ```suggestion
Migrate the code related to your component's domain from `pkg/flare` to your component. After migrating
```

### Creating a callback

> To add data to a flare, register a provider.

Author: Clearer ?
```suggestion
To add data to a flare you will fiest need to register a `FlareBuilder`.
```

> of the Agent to components we should be able to delete `pkg/flare`.

Reviewer: ```suggestion
the Agent to components, you can delete `pkg/flare`.
```

> ### Creating you callback

Reviewer: ```suggestion
### Creating a callback
```

> In order to add data to a flare you need to register a provider.

Reviewer: ```suggestion
To add data to a flare, register a provider.
```

> First create a `flare.go` file in your component (the file name is a convention) and create a `func (c *yourComp) fillFlare(fb flarehelpers.FlareBuilder) error` function.

Author: I think it makes it much easier to work with a feature across multiple component. I was thinking about cases where we want to troubleshoot some behavior, that way we don't "struggle" to find the right piece of code in charge of `flare`, `health` or `status page` in a component.

I was thinking we could expand this to all cross-component feature. Each component would have a `heath.go`, `status_page.go` or `flare.go`. 

Definitely open to suggestion here, but I thought it would be useful.

For the method name, it was to ease searching thought the code. But same as before, open to debate :man_shrugging: .

> First create a `flare.go` file in your component (the file name is a convention) and create a `func (c *yourComp) fillFlare(fb flarehelpers.FlareBuilder) error` function.

Reviewer 2: Searching for `flarehelpers.NewProvider` gives all the entry points of the flare. Let's discuss it offline.

> First create a `flare.go` file in your component (the file name is a convention) and create a `func (c *yourComp) fillFlare(fb flarehelpers.FlareBuilder) error` function.

Author: After discussing: Removing the convention for now, it might not be that useful. We will come back to the idea of convention for cross component feature in the future.

First create a `flare.go` file in your component (the file name is a convention) and create a `func (c *yourComp) fillFlare(fb flarehelpers.FlareBuilder) error` function.

> This function will be called everytime the agent generate a flare, either from the CLI or from the running agent. This
> callback will receive a `comp/core/flare/helpers:FlareBuilder`. The `FlareBuilder` is an interface that provides all the
> required helpers needed to add data to a flare (add files, copy directory, scrub data, ...).

Reviewer: ```suggestion
This function is called every time the Agent generates a flare, either from the CLI or from the running Agent. This
callback receives a `comp/core/flare/helpers:FlareBuilder`. The `FlareBuilder` interface provides the
helpers required to add data to a flare: adding files, copying directories, scrubbing data, and so on.
```

Example of `flare.go`:

```golang
import (
	yaml "gopkg.in/yaml.v2"

	flarehelpers "github.com/DataDog/datadog-agent/comp/core/flare/helpers"
)

func (c *cfg) fillFlare(fb flarehelpers.FlareBuilder) error {
	fb.AddFileFromFunc(
		"runtime_config_dump.yaml",
		func () ([]byte, error) {
			return yaml.Marshal(c.AllSettings())
		},
	)

	fb.CopyFile("/etc/datadog-agent/datadog.yaml")
	return nil
}
```

> Read the package documentation for `FlareBuilder` for more information. But note that all errors will automatically be
> added to a log file shipped within the flare. You should try to ship as much data as possible in a flare instead of
> stopping at the first error. Returning an error will not stop the flare from being created nor sent.

Reviewer: ```suggestion
Read the package documentation for `FlareBuilder` for more information. 
All errors are automatically 
added to a log file shipped within the flare. Ship as much data as possible in a flare instead of
stopping at the first error. Returning an error does not stop the flare from being created or sent.
```

### Migrating your code

> The code in `pkg/flare` already use the `FlareBuilder`, so migration should be simple. You need to locate the code
> related to your component domain from `pkg/flare` and move it to your `fillFlare` function.

Reviewer: ```suggestion
The code in `pkg/flare` uses the `FlareBuilder` interface, simplifying migration. Locate the code
related to your component domain from `pkg/flare` and move it to your `fillFlare` function.
```

### Register the callback

> Finally, to Register your callback, provide a new `comp/core/flare/helpers:Provider` by using `comp/core/flare/helpers:NewProvider`.

Author: I agree, I updated the doc, tell me what you think.

> ### Register you callback

Reviewer: ```suggestion
### Register the callback
```

> Finally, to Register you callback you need to provide a new `comp/core/flare/helpers:Provider`. The function `comp/core/flare/helpers:NewProvider` is there for that.

Reviewer: ```suggestion
Finally, to Register your callback, provide a new `comp/core/flare/helpers:Provider` by using `comp/core/flare/helpers:NewProvider`.
```
